source/common/http/conn_pool_base.cc

```diff
@@ -40,7 +40,16 @@ void ConnPoolImplBase::purgePendingRequests(
 
 void ConnPoolImplBase::onPendingRequestCancel(PendingRequest& request) {
   ENVOY_LOG(debug, "cancelling pending request");
-  request.removeFromList(pending_requests_);
+  if (!pending_requests_to_purge_.empty()) {
+    // If pending_requests_to_purge_ is not empty, it means that we are called from
+    // with-in a onPoolFailure callback invoked in purgePendingRequests (i.e. purgePendingRequests
+    // is down in the call stack). Remove this request from the list as it is cancelled,
+    // and there is no need to call its onPoolFailure callback.
+    request.removeFromList(pending_requests_to_purge_);
```

Comment:
I just realized there might be a (admittedly contrived) problem with this code. What if a user adds a request to the connection pool, and then immediately cancels it? I'm pretty sure this will break. I can't imagine why anyone would do this but it seems like we should at least comment/assert this somehow? WDYT?

Reply:
If I understand correctly, this scenario will only happen if you cancel a request from the cancellation callback of that same request. I can add a comment to `Callbacks::onPoolFailure` to warn the user to not call `cancel()` in that case. Does that sound good?

as far as assert, I believe this will be covered by the existing assert; i.e. someone who tries this will encounter a stack trace similar to the one I posted (with or without this change).

Reply:
I can't see anyone ever doing this, but it's worth adding a comment somewhere about this.

@alyssawilk @snowp any thoughts on this?

/wait